src/plugin.ts

```diff
@@ -0,0 +1,212 @@
+}
+
+interface IPluginIFrameSettings {
+  readonly highlighterEnabled: boolean;
+}
+
+export interface IPluginConfiguration {
+  readonly languageCodename: string | null;
+  readonly projectId: string | null;
+  readonly queryParam: string;
+}
+
+const initialState: IPluginState = {
+  queryTimerId: 0,
+};
+
+const defaultConfiguration: IPluginConfiguration = {
+  languageCodename: null,
+  projectId: null,
+  queryParam: 'kk-plugin-enabled',
+};
+
+class Plugin {
+  private state: IPluginState;
+  private configuration: IPluginConfiguration;
+  private iFrameCommunicator?: IFrameCommunicator;
+  private readonly nodeHighlighter: NodeHighlighter;
+
+  constructor(configuration?: Partial<IPluginConfiguration>) {
+    this.nodeHighlighter = new NodeHighlighter();
+
+    this.state = initialState;
+    this.configuration = { ...defaultConfiguration, ...configuration };
+
+    if (isInsideIFrame()) {
+      this.initializeIFrameCommunication();
+    } else {
+      this.watchQueryParams();
+    }
+
+    this.listenToHighlighterEvents();
+  }
+
+  public destroy = (): void => {
+    this.unwatchQueryParams();
+    this.nodeHighlighter.destroy();
+    this.iFrameCommunicator?.destroy();
+  };
+
+  public updateConfiguration = (configuration: Partial<IPluginConfiguration>): void => {
+    this.configuration = { ...(this.configuration || defaultConfiguration), ...configuration };
+  };
+
+  private updateState = (state: Partial<IPluginState>): void => {
+    this.state = { ...this.state, ...state };
+  };
+
+  private initializeIFrameCommunication = (): void => {
+    if (!isInsideIFrame() || this.iFrameCommunicator) return;
+
+    this.iFrameCommunicator = new IFrameCommunicator();
+
+    const storage = createStorage<IPluginIFrameSettings>('kk:plugin:iframe-settings');
+    const stored = storage.get();
+    const highlighterEnabled = stored !== null ? stored?.highlighterEnabled : true;
+
+    if (highlighterEnabled) {
+      this.nodeHighlighter.enable();
+    }
+
+    this.iFrameCommunicator.addMessageListener(
+      IFrameMessageType.HighlightsStatus,
+      (data: IHighlightsStatusMessageData) => {
+        if (!data || !this.nodeHighlighter) return;
+
+        if (data.enabled) {
+          this.nodeHighlighter.enable();
+        } else {
+          this.nodeHighlighter.disable();
+        }
+
+        storage.set({
+          highlighterEnabled: data.enabled,
+        });
+      }
+    );
+
+    this.iFrameCommunicator.sendMessage(IFrameMessageType.Initialized, {
+      projectId: this.configuration.projectId,
+      languageCodename: this.configuration.languageCodename,
+      highlighterEnabled,
+    });
+  };
+
+  private watchQueryParams = (): void => {
+    if (isInsideIFrame()) return;
+
+    if (this.state.queryTimerId) {
+      clearTimeout(this.state.queryTimerId);
+    }
+
+    const currentPageUrl = new URL(document.URL);
+    const isPluginEnabled = currentPageUrl.search.includes(this.configuration.queryParam);
```

Comment:
This will enable the plugin even in case I include `?videoPreview=false` for other module or `utm_source=preview` or other totally unrelated parameters. The check should be probably more robust